### ISSUE TITLE
Provide lint that warns .. in struct expression has no effect

### ISSUE THREAD

**Issue by frewsxcv** [rust-lang/rust#23199]
As an example:

```rust
struct MyStruct {
    pub a: i64,
}

fn main() {
    let existing = MyStruct {a: 50};
    MyStruct {
        a: 100,
        ..existing
    };
}
```

The line containing `..existing` is unnecessary since we already explicitly state all the fields of the struct. Maybe we could have a lint so that something like this gets shown:

```
tmp.rs:12:17: 12:38 warning: this statement has no effect
tmp.rs:12                    ..existing
                             ^~~~~~~~~~
```

**Comment by Thiez** [comment]
I can't really think of a scenario in which this would actually lead to a bug, and it could be useful to do such a useless `..existing` if you intend to add more fields to the struct in the future. Can you think of any bad behavior that the proposed lint would prevent?

**Comment by frewsxcv** [comment]
Sorry to answer your question with a question: What bad behavior does warning about unused imports prevent? I thought the lints were prevent against bad behaviors *in addition to* promoting simpler/cleaner code. I don't feel too strongly about my proposal; I just thought of it recently and wanted to know what others thought.

**Comment by pnkfelix** [comment]
@frewsxcv are you planning for the lint to only fire if the `..existing` expression has no side-effects?  (perhaps approximated by only having it fire for trivial expressions like identifiers)?  Just curious.

(I personally think the compiler err's on the side of including too many lints; the unnecessary parens and unused imports lints immediately come to my mind.  But then again, one is nearly always free to turn off lints, so I don't fight the trend too hard.)

**Comment by frewsxcv** [comment]
> are you planning for the lint to only fire if the ..existing expression has no side-effects? 

Basically, yes

**Comment by kmcallister** [comment]
I think this belongs in a third-party library like [rust-clippy](https://github.com/Manishearth/rust-clippy).  If it proves useful then we can consider merging it at a later date.

**Comment by sanxiyn** [comment]
This has been implemented in rust-clippy.
